### ISSUE TITLE
Allow configuring the gpg program used for signing & verification

### ISSUE THREAD

**Issue by gregoire-mullvad** [containers/podman#15838]
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind feature

**Description**

I would like to be able to sign containers using podman, but using a specific gpg binary to do the actual signing.  Something similar to git's `gpg.program` config.  (More specifically, I would like to use Qubes OS [Split GPG setup](https://www.qubes-os.org/doc/split-gpg/).)

I dug around a bit and I believe this can be done by allowing the user to overwrite the second argument of `SetEngineInfo` in [github.com/containers/image/signature](https://github.com/containers/image/blob/2bb3f3e44c5cbe532a3cdbe735f78c21418f8dc2/signature/mechanism_gpgme.go#L81).  I should probably open an issue on containers/image, but I wanted to check first that it would be ok to expose this through podman somehow (option, config, env var…)

**Comment by mheon** [comment]
@mtrmac PTAL

**Comment by rhatdan** [comment]
I would have no problem with it.

**Comment by mtrmac** [comment]
I’m fine with that in principle. 

Just noting that it would be a fairly extensive change in c/image, most of the policy code is not currently carrying around a `types.SystemContext`. (And at this point I’m not sure if we want to do that, or design an start using some very different option mechanism.)

**Comment by rhatdan** [comment]
@gregoire-mullvad Interested in opening a PR?

**Comment by gregoire-mullvad** [comment]
I could give it a go, but most likely not before one or two weeks.

**Comment by rhatdan** [comment]
@gregoire-mullvad Did you ever get a chance?

**Comment by rhatdan** [comment]
@mtrmac @vrothberg Would it make more sense to define the gpg program inside of registries.conf rather then allowing users on the fly to choose one.  I don't think this will be variable per container, more of a system wide setting.

**Comment by mtrmac** [comment]
`registries.conf` is clearly not the right file. Not that we _have_ the right file…
- `policy.json` is conceptually close, but “policy” is not a great fit for implementation details like this (especially when c/image can be built so that it does not use GPG at all; would it ignore that setting, or fail?)
- `containers.conf` is general enough, but not used by Skopeo. Still, I guess it’s the closest match.

(Actually building a compatible API extension throughout c/image to allow passing that configuration would be a the bigger part of the work.)

**Comment by gregoire-mullvad** [comment]
I did start looking into this a bit. My initial ideas was to duplicate [`NewGPGSigningMechanism()`](https://github.com/containers/image/blob/2bb3f3e44c5cbe532a3cdbe735f78c21418f8dc2/signature/mechanism.go#L59) into something like `NewGPGSigningMechanismWithProgram(gpgProgram string)`. @mtrmac do you think that would be an acceptable way to pass this configuration to c/image?
(I considered something like `NewGPGSigningMechanismWithOpts` which would take a struct that can later be extended, but it seems that there aren't a lot more options can be passed to gpgme, and if the user can override the gpg binary they can write a wrapper that sets whatever option they want anyway.)

As for the user interface, I was thinking of just adding a new `--gpg-program` flag to `podman-image-sign` to start with.

**Comment by mtrmac** [comment]
> I did start looking into this a bit. My initial ideas was to duplicate [`NewGPGSigningMechanism()`](https://github.com/containers/image/blob/2bb3f3e44c5cbe532a3cdbe735f78c21418f8dc2/signature/mechanism.go#L59) into something like `NewGPGSigningMechanismWithProgram(gpgProgram string)`.

We would also want that for `newEphemeralGPGSigningMechanism` for consistency; and then the value needs to propagate through `prSignedBy.isSignatureAuthorAccepted`, i.e. probably be somehow configurable when creating a `PolicyContext`. So that, and the `copy.copier.createSignature` path, both needs to pass through GPG options without really caring what they are — and that suggests to me that we do need a struct, not just so that it can be extended, but also so that the large call stack doesn’t need to manage items individually.

And the generic `PolicyContext`, which isn’t really GPG-exclusive, should probably define the configuration mechanism of the new variant of `NewPolicyContext` in a way that can be extended for other implementation choices, so not just a `NewPolicyContextWithGPGOptions`.

---

(I could see an argument for adding the option only to `NewGPGSigningMechanism`, not the ephemeral one, on the theory that for signing it’s more important to have a choice of how to manage the private keys, while the ephemeral verifiers don’t strictly need to allow users’ implementation choice. I don’t quite agree but I don’t feel strongly about it.)

**Comment by gregoire-mullvad** [comment]
I thought a bit more about this, and I think it's not obvious that the ephemeral mechanism should be changed for consistency. At least in the use case I described above, gpg-split in Qubes, the "alternative" gpg program will ignore `GNUPGHOME` by design. So it might be dangerous to try to set `GNUPGHOME` while explicitly allowing the user to chose the gpg program. And in that case, maybe it makes sense to only allow changing the program in the non-ephemeral case and only add the configuration option for signing, and not for verification?

**Comment by rhatdan** [comment]
Conversation on this seems to have died out, I just saw it in the stale-issues list.
@gregoire-mullvad is this something you are still interested in working on?

**Comment by gregoire-mullvad** [comment]
Yes. I can try to open a PR on containers/image this or next week.
